### ISSUE TITLE
Problem consuming stream with a mixed model table

### ISSUE THREAD

**Issue by Arttii** [numberoverzero/bloop#110]
Hi,

As far as I understood bloop fully supports using 2 different models on the same table, with some fields missing in one model from the other. 
I could not figure out how to consume from dynamodb streams with multiple models. I just get exceptions on deserializing message from one model when using the other one.

Is this somehow a thing we should consider or is it an omission?

Thanks

**Comment by numberoverzero** [comment]
Thanks for opening an issue.  This is a good question, and one I hadn't explored much.

Your models may overlap in a few ways; do they share any columns?  Do those columns have the same types?  Are you using any custom types?  Do those types handle `None` correctly?  It may be easier to answer your question if you can provide sanitized models that demonstrate the problem.

Below I've tried to answer the general question of consuming 1 table: N models in a single Stream.  Let me know if that's off the mark.

# tl;dr

The short answer is that `Stream` won't handle this for you today, but you could build this pretty easily on top of the streaming internals.

I'd consider adding a `Stream` overload (or helper function) that takes a user-provided `record -> model class` function to pick the correct model, and does everything else the current Stream class does.  I'd expect that signature to look something like (naming tbd):

```python
TableStream(engine, stream_arn, model_selector)
```

and used as follows:

```python
def pick_model(record):
    if record["type"] == {"S": "e"}:
        return Employee
    else:
        return Manager

stream = TableStream(engine, Employee.meta.stream["arn"], pick_model)
while True:
    record = next(stream)
    if not record: continue
    print(type(record["new"]))
```

# Background

Today, `Stream` must know which model to create instances of; it doesn't use a discriminator in the record itself to find the correct model.  Without the user inspecting the record (and considering their calling context) it's not possible to know which model they expect to use.  Baking that logic into the class would result in most users either not using Streams, or copying some 1:1 model function from the Patterns section every time they wanted to consume a stream.  I wanted to optimize for the 1:1 model:table case, which I expect accounts for well over 99% of all DynamoDB users.

As a quick example, we have these shared models:

```python
class Employee(BaseModel):
    class Meta:
        table_name = "employees-uk"
    id = Column(UUID, hash_key=True)
    level = Column(Integer)
    email = Column(String)
    manager_id = Column(UUID)

class Manager(BaseModel):
    class Meta:
        table_name = "employees-uk"
    id = Column(UUID, hash_key=True)
    level = Column(Integer)
    email = Column(String)
    manager_id = Column(UUID)
    directs = Column(Set(UUID))
```

And this record comes in.  Is this a new `Employee` or simply a `Manager` with no directs?
```
{
  id: {"S": "6d8b54a2-fa07-47e1-9305-717699459293"},
  level: {"N": "4"},
  email: {"S": "user@domain.com"},
  manager_id: {"S": "2a45b8d6-fa07-47e1-9305-717699459293"}
}
```

Attaching a "type" column with "employee" or "manager" isn't enough either, because some users may want to use different models based on the "created_at" time (they know OldModel was used before June 2016 and NewModel was used after June 2016).  The user will always need to tell bloop which model to use for the record; currently it's effectively a hardcoded function with the model you pass in.

# Building your own

While not public, you can use `bloop.stream.coordinator.Coordinator` to consume the records without unpacking them into models, and apply your own logic to create models using `bloop.models.unpack_from_dynamodb`.  That's all the `Stream` class does; wraps a coordinator and unpacks into model objects.

* [`bloop.Stream.__next__` fetching a record from the coordinator and unpacking it](https://github.com/numberoverzero/bloop/blob/81d0984798749a674178f342e5cf96ae80b48f8b/bloop/stream/stream.py#L28-L37)  (note that each of the "new", "old" and "key" dicts are unpacked individually; see [Record Structure](http://bloop.readthedocs.io/en/latest/user/streams.html#record-structure) for more details)
* [`bloop.models.unpack_from_dynamodb` takes in a single record and returns a new instance](https://github.com/numberoverzero/bloop/blob/81d0984798749a674178f342e5cf96ae80b48f8b/bloop/models.py#L581-L598)
* Issue #84 may be related, and together with exposing the `Coordinator` could be enough to cover your needs.

This is probably where I'd start for the `TableStream` described above.  It's a quick modification from the actual `Stream` class, with comments stripped (**and untested**): https://gist.github.com/numberoverzero/01eef671aeda8c42bf9dcbe3ddc76321

**Comment by Arttii** [comment]
Hey, thanks for the very detailed answer :D.

I was thinking exactly that, and kinda hacked around the Streams class to achieve something similar.
I like having a selector function, I was thinking more in the realm of of specifying all the models in a list, but a function should be much more portable and non magic.

I tested your implementation, I just had to add the `move_to` position call:

```python
class TableStream:
    def __init__(self, *, stream_arn, engine, model_selector,position):
        """
        model_selector is a function that takes a record
        and returns a model class.  for example:
            def model_selector(record):
                if record["new"]["bloop_type"] == {"S": "employee"}:
                    return Employee
                else:
                    return Manager
        """
        self.engine = engine
        self.coordinator = Coordinator(
            session=engine.session,
            stream_arn=stream_arn)
        self.model_selector = model_selector
        self.coordinator.move_to(position)
```
I tested with the following models:
```python
class CatalogItemBase(BaseModel):
    class Meta:
        abstract = True

    place_id = Column(UUID, hash_key=True)

    type_item_id = Column(String, range_key=True)

    item_id = Column(UUID, default=uuid.uuid4)
    type = Column(StringEnum(CatalogItemType))

    name = Column(String)
    description = Column(String)

    created_at = Column(DateTime, default=maya.now().datetime)
    updated_at = Column(DateTime, default=maya.now().datetime)


class CatalogItem(CatalogItemBase):
    class Meta:
        table_name = config.CATALOG_TABLE
        stream = {
            "include": ["new","old"]
        }

    CatalogItemCustomization = Map(**{
        "customization_id": UUID,
        "limit": Integer
    })


    price = Column(Integer)
    category = Column(StringEnum(CatalogItemCategory), default=CatalogItemCategory.MAIN)

    customizations = Column(List(CatalogItemCustomization))


class CatalogItemCustomization(CatalogItemBase):
    class Meta:
        table_name = config.CATALOG_TABLE
        stream = {
            "include": ["new","old"]
        }

    CatalogItemMap = Map(**{
        "item_id": UUID,
        "name": String,
        "description":String,
        "price": Integer
    })

    limit = Column(Integer)

    customization_items = Column(List(CatalogItemMap))
```
And the code to init the streams:
```python
def model_selector(record):
    if record["new"]["type"] == {"S": "CUSTOMIZATION"}:
        return CatalogItemCustomization
    else:
        return CatalogItem

#items = engine.stream(CatalogItemCustomization, "trim_horizon")
items =TableStream(stream_arn=CatalogItem.Meta.stream.get("arn"),engine=engine,model_selector=model_selector,position="trim_horizon")
```
Seems to work quite ok. I agree its not a primary usecase for people to consume  streams of mixed model tables, but the thing is, seeing how its quite supported in bloop(whichIi find really cool), its annoying to have to handle the exceptions coming from the `_unpack`, especially as I have some validation going on the Columns.

So I think this functionality has some merit and could be useful.

Feel free to close the issue, or keep it open as a tracking issue. I might work on a PR if you're not already done before I even start :D.

**Comment by numberoverzero** [comment]
Thanks for the detailed sample!

Given those two models, I'm not sure I understand what was causing the original problem.  I would expect a table with those two models to deserialize without throwing an exception, although there might be missing values: for example, if you streamed as `CatalogItem` and encountered a row that was actually a `CatalogItemCustomization` then you wouldn't see the `category` column filled.

Can you provide an example of the exception you were hitting in `_unpack`?  I'm especially surprised because (except for the shared columns in `CatalogItemBase`) your column names don't even overlap, so it shouldn't be due to trying to deserialize a column with the wrong type.

**Comment by Arttii** [comment]
The main issue is that I am using the modified Column class with validation from the bloop patterns and its creating some problems. The thing is, I still think its better to be explicit and not just assume None means we are probably getting values from one Model or the other. We would basically have to check the type field anyway to use any business logic and then apply `CatalogItemCustomization` on an instance of `CatalogItem`. Then again, I might be overcomplicating things.

**Comment by numberoverzero** [comment]
Hey @Arttii, I'm going to close this for now.  If there's renewed interest in a more dynamic stream interface (specifically: not locked to one model at a time) I'm happy to re-open.

Thanks again for trying out some of the newer features of bloop! :rainbow:
